### DISCUSSION TITLE
"Please install the missing version" after adding new Python version in new project

### DISCUSSION THREAD

**Discussion by BillyUdders** [asdf-community/asdf-direnv#54] [Q&A] [answered]
**Description**

I've been using asdf-direnv for a few weeks in a Python project that uses Python 3.6. Python version has been installed with `asdf install python 3.6.10`. Now I've added a new version of python 3.8.3 in the same way. In my new project I've defined the .envrc file below. When I `asdf local python 3.8.3` in my new project and then `asdf exec direnv allow .envrc` I get the error below.

1) Install a version of asdf'd python and setup a direnv'd project with it
2) Install and setup another version of python, make a project and set it's version locally in .tool-version
3) asdf exec direnv allow .envrc or cd to allowed new project directory
4) 126 error
5) Notice that the version of `python -V` in new project directory is `3.8.3`
6) Notice that 126 error stops a .direnv being created.

I tried reinstalling the plugin because I installed one version of Python before installing the asdf-direnv plugin and one after but nothing seems to have changed.

**OSX and stuff:**

MacOS Catalina
Fish Shell
asdf installed via Brew

**Full error message, after calling asdf exec direnv allow .envrc:**

```
direnv: loading .envrc
direnv: using asdf
direnv: loading ~/.asdf/installs/direnv/2.20.0/env/2727433909-3802552236-3986344091-3702538069
direnv: using asdf terraform 0.11.1
direnv: using asdf direnv 2.20.0
direnv: using asdf python 3.8.3
asdf: No preset version installed for command virtualenv
Please install the missing version by running one of the following:

asdf install python 3.8.3

or add one of the following in your .tool-versions file:

python 3.6.10
direnv: error exit status 126
```

**.envrc file:**

```
use asdf
layout python3
```

**direnvrc**

```
source $(asdf which direnv_use_asdf)


# Uncomment the following line to make direnv silent by default.
# export DIRENV_LOG_FORMAT=""
```

**Comment by michi-zuri** [comment]
Two things to check: Try without the line `layout python3` in your .envrc to see what happens and secondly, see if you can reproduce with the latest release of asdf-direnv. In my experience installing Python through miniconda is easier on macOS, so you could also give that a try.

**Comment by cho-m** [comment, marked as answer]
You probably already resolved this since it has been a while, but I'll note that `virtualenv` is a pip-installed Python package accessible via asdf shim and the error message is saying that the package isn't installed for your Python 3.8.3.

If you upgraded direnv from 2.20.0 to a newer version, I think the issue should have gone away since direnv switched to prioritize stdlib `venv` over `virtualenv` in 2.21.0: https://github.com/direnv/direnv/blob/master/CHANGELOG.md#stdlib

> `layout_python`: prefer venv over virtualenv. Do not export VIRUAL_ENV if $python_version is unavailable or a virtual environment does not exist/can't be created

**Reply by jfly** [reply]
Another solution would be to get your asdf shims directory out of your PATH, and then you won't have this weird `virtualenv` wrapper script that doesn't know what to do. See our [pro tips](https://github.com/asdf-community/asdf-direnv#pro-tips) for instructions on how to do that.
